lib/publiccloud/provider.pm

```diff
@@ -126,6 +121,94 @@ sub parse_ipa_output {
     return $ret;
 }
 
+=head2 create_ssh_key
 
-1;
+Creates an ssh keypair in a given file path by $args{ssh_private_key_file}
+
+=cut
+sub create_ssh_key {
+    my ($self, %args) = @_;
+    if (script_run('test -f ' . $args{ssh_private_key_file}) != 0) {
+        assert_script_run('SSH_DIR=`dirname ' . $args{ssh_private_key_file} . '`; mkdir -p $SSH_DIR');
+        assert_script_run('ssh-keygen -b 2048 -t rsa -q -N "" -f ' . $args{ssh_private_key_file});
+    }
+}
+
+=head2 run_ipa
+
+called by childs within ipa function
+
+=cut
+sub run_ipa {
+    my ($self, %args) = @_;
+    $args{cleanup}              //= 1;
+    $args{ssh_private_key_file} //= '.ssh/id_rsa';
+    $args{tests}                //= '';
+    $args{timeout}              //= 60 * 20;
+    $args{results_dir}          //= 'ipa_results';
+    $args{distro}               //= 'sles';
+    $args{tests} =~ s/,/ /g;
+
+    my $provider = get_var('PUBLIC_CLOUD_PROVIDER');
```

Comment:
I would prefer using `$args{provider}`. And maybe also die if it's not present in `%args`.

Reply:
already changed in latest commit :)